### PR TITLE
[GEOS-10658] - Added OpenLayers example using vectortiles in WMS for vector tiles tutorial

### PR DESCRIPTION
**Pull request by ale-cristofori**
[GEOS-10658](https://osgeo-org.atlassian.net/browse/GEOS-10658)

This PR contains integrations in Geoserver docs to include an example for creating an OpenLayers map using Geoserver, Openlayers with a on overlay vectortiles based layer using WMS 
  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [ ] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [ ] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)). N/A in this pr

For core and extension modules:

- [ ] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [ ] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [ ] Bug fixes and small new features are presented as a single commit.
- [ ] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Comment by ale-cristofori** [comment]
@aaime, I was told you'd be the best person to review this. I have no permission to change the assignee on this PR. Could you assign this to yourself? Thanks

File: doc/en/user/source/extensions/vectortiles/tutorial.rst [hunk 1]
```diff
@@ -46,7 +46,7 @@ GeoServer can also produce vector tiles in three formats: GeoJSON, TopoJSON, and
      - This is a human readable JSON format.  Although many geo-spatial applications support GeoJSON datasets, few Vector Tile applications support tiles in this format.  Supported by Open Layers 3.
    * - `TopoJSON <https://github.com/mbostock/topojson/wiki>`_
      - ``application/json;type=topojson``
-     - This is a very complex, but somewhat human readable JSON format that is good for polygon coverages.  It is not a widely supported and very few Vector Tile applications support it.  Supported by Open Layers 3.
+     - This is a very complex, but somewhat human readable JSON format that is good for polygon coverages.  It is not a widely supported and very few Vector Tile applications support it.  Suported by Open Layers 3.
```
Comment (bradh) [inline doc/en/user/source/extensions/vectortiles/tutorial.rst:49]: Why this change?

File: doc/en/user/source/extensions/vectortiles/tutorial.rst [hunk 2]
```diff
@@ -27,7 +27,7 @@ Vector tile formats
 
 GeoServer can also produce vector tiles in three formats: GeoJSON, TopoJSON, and MapBox Vector (MVT). These are also supported by OpenLayers and other clients.
 
-.. warning:: When using vector tiles, be sure to use an up-to-date client. Older clients do not support all vector tiles capabilities and may result in rendering errors. We recommend using the latest version of OpenLayers (Currently v5.3.0).
+.. warning:: When using vector tiles, be sure to use an up-to-date client. Older clients do not support all vector tiles capabilites and may result in rendering errors. We recommend using the latest version of OpenLayers (Currently v5.3.0).
```
Comment (bradh) [inline doc/en/user/source/extensions/vectortiles/tutorial.rst:30]: Why this change?

File: doc/en/user/source/extensions/vectortiles/tutorial.rst [hunk 3]
```diff
@@ -169,6 +169,115 @@ Create OpenLayers application
 
 These tiles are being rendered by the OpenLayers client.
 
+Create OpenLayers application - WMS Vector Tiles
+------------------------------------------------
+
+.. note::   Vector tiles requested with WMS allows to retrieve non-cached vector tiles (server side), setting the ``tiled=false`` parameter on the ``getMap`` request. This setting could be particularly useful when serving fast changing source data that should constantly be kept up-to-date for display.
```
Comment (bradh) [inline doc/en/user/source/extensions/vectortiles/tutorial.rst:175]:
```suggestion
.. note::   Vector tiles requested with WMS allows retrieving non-cached vector tiles (server side) by setting the ``tiled=false`` parameter on the ``getMap`` request. This setting could be particularly useful when serving fast changing source data that should constantly be kept up-to-date for display.
```

File: doc/en/user/source/extensions/vectortiles/tutorial.rst [hunk 4]
```diff
@@ -169,6 +169,115 @@ Create OpenLayers application
...
+
+#. Create a ``www/wms-vectortiles`` directory inside your GeoServer Data Directory.
+
+#. Download the `latest version of OpenLayers <http://openlayers.org/download/>`_.
+
+#. Extract the following files to from the downloaded archive to the directory created in step 1:
```
Comment (bradh) [inline doc/en/user/source/extensions/vectortiles/tutorial.rst:181]:
```suggestion
#. Extract the following files from the downloaded archive to the directory created in step 1:
```

File: doc/en/user/source/extensions/vectortiles/tutorial.rst [hunk 5]
```diff
@@ -169,6 +169,115 @@ Create OpenLayers application
...
+            font-family: sans-serif;
+            width: 100%;
+          }
+          .map {
+            height: 500px;
+            width: 100%;
+          }
+        </style>
+      </head>
+        <body>
```
Comment (bradh) [inline doc/en/user/source/extensions/vectortiles/tutorial.rst:209]: Looks like the formatting is off. Possibly reformat it before incorporating it?

File: doc/en/user/source/extensions/vectortiles/tutorial.rst [hunk 6]
```diff
@@ -80,10 +80,10 @@ For this tutorial, we'll be using the layer ``opengeo:countries`` to show off th
 
 Our layer is now ready to be served.
 
-Create OpenLayers application
------------------------------
+Create OpenLayers application - TMS Vector Tiles
+------------------------------------------------
 
-#. Create a ``www/vectortiles`` directory inside your GeoServer Data Directory.
+#. Create a ``www/tms-vectortiles`` directory inside your GeoServer Data Directory.
```
Comment (bradh) [inline doc/en/user/source/extensions/vectortiles/tutorial.rst:86]: Why this change?

**Review by bradh — CHANGES_REQUESTED** [review]

File: doc/en/user/source/extensions/vectortiles/tutorial.rst [hunk 7]
```diff
@@ -169,6 +169,115 @@ Create OpenLayers application
...
 These tiles are being rendered by the OpenLayers client.
 
+Create OpenLayers application - WMS Vector Tiles
+------------------------------------------------
+
+.. note::   Vector tiles requested with WMS allows to retrieve non-cached vector tiles (server side), setting the ``tiled=false`` parameter on the ``getMap`` request. This setting could be particularly useful when serving fast changing source data that should constantly be kept up-to-date for display.
+
+#. Create a ``www/wms-vectortiles`` directory inside your GeoServer Data Directory.
+
+#. Download the `latest version of OpenLayers <http://openlayers.org/download/>`_.
```
Comment (aaime) [inline doc/en/user/source/extensions/vectortiles/tutorial.rst:179]: I see two downloads, site and package... I guess you mean the "package" one?
I see "ol.css" in the root of the zip, "ol.js" in the "dist" zip, don't see a ``ol-debug.js`` though.
This part of the instructions should contain a bit more detail.

File: doc/en/user/source/extensions/vectortiles/tutorial.rst [hunk 8]
```diff
@@ -169,6 +169,115 @@ Create OpenLayers application
...
+      </head>
+        <body>
+          <h3>Mapbox Protobuf - vector tiles WMS</h3>
+          <div class="refresh-container">
+          <button id="refresh-button" type="button" onclick="updateFunc();">Refresh/reload cache</button>
+          </div>
+          <div id="map" class="map"></div>
+          <script>
+          
+            var layerParams = {'LAYERS': 'opengeo:countries', 'TILED': false, 'format': 'application/vnd.mapbox-vector-tile'};
```
Comment (aaime) [inline doc/en/user/source/extensions/vectortiles/tutorial.rst:217]: Something odd here, the generate URL ends up having "format" twice, once uppercase, once lowercase, e.g:

```
http://localhost:8080/geoserver/wms?SERVICE=WMS&VERSION=1.3.0&REQUEST=GetMap&FORMAT=image%2Fpng&TRANSPARENT=true&LAYERS=importer%3Ane_10m_roads&TILED=false&format=application%2Fvnd.mapbox-vector-tile&WIDTH=256&HEIGHT=256&CRS=EPSG%3A3857&STYLES=&BBOX=10018754.17139462%2C0%2C20037508.34278924%2C10018754.171394622
``

**Review by aaime — CHANGES_REQUESTED** [review]

File: doc/en/user/source/extensions/vectortiles/tutorial.rst (hunk 7 above)
Comment (ale-cristofori) [reply doc/en/user/source/extensions/vectortiles/tutorial.rst:179]: @aaime, yes, I copied and pasted it from the previous part. Both should be rewritten since the downloadable `.zip` of OpenLayers has changed structure

File: doc/en/user/source/extensions/vectortiles/tutorial.rst (hunk 5 above)
Comment (ale-cristofori) [reply doc/en/user/source/extensions/vectortiles/tutorial.rst:209]: yes it is off

File: doc/en/user/source/extensions/vectortiles/tutorial.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (aaime) [reply doc/en/user/source/extensions/vectortiles/tutorial.rst]: It should really be qualified by the size of the vector tile generated. One needs to pay attention to generate small ones (few attributes, few features) in order for them to be usable.

Also editing on vector tiles is problematic, geometry locations are quantized on a grid, you are never editing the original position, but a "close" one, how close depends on your current zoom level, just mind it will never be exactly the original one. Also, if dealing with lines/polygons crossing the tile borders, they are also clipped, reconstructing them on the client side from all the tiles containing them (good luck finding when to stop, think the coastline of Asia for example) is possible but causes further accuracy degradation.

File: doc/en/user/source/extensions/vectortiles/tutorial.rst (hunk 1 above)
Comment (ale-cristofori) [reply doc/en/user/source/extensions/vectortiles/tutorial.rst:49]: I have to be fair I am not too sure what happened there. I fixed it anyway

File: doc/en/user/source/extensions/vectortiles/tutorial.rst (hunk 2 above)
Comment (ale-cristofori) [reply doc/en/user/source/extensions/vectortiles/tutorial.rst:30]: I have to be fair I am not too sure what happened there. I fixed it anyway

[the remark this answers is not in the document]
Comment (ale-cristofori) [reply doc/en/user/source/extensions/vectortiles/tutorial.rst:30]: Replied above to this

File: doc/en/user/source/extensions/vectortiles/tutorial.rst (hunk 7 above)
Comment (aaime) [reply doc/en/user/source/extensions/vectortiles/tutorial.rst:179]: Then please do so :-D

File: doc/en/user/source/extensions/vectortiles/tutorial.rst (hunk 6 above)
Comment (ale-cristofori) [reply doc/en/user/source/extensions/vectortiles/tutorial.rst:86]: @bradh, this was to differentiate this example from the one I included. In case users want to follow both and have them in separated folders `tms-vectortiles` `wms-vectortiles`.

File: doc/en/user/source/extensions/vectortiles/tutorial.rst (hunk 8 above)
Comment (ale-cristofori) [reply doc/en/user/source/extensions/vectortiles/tutorial.rst]: @aaime yes 'FORMAT' had to be specified in upper case for the property to be overridden

File: doc/en/user/source/extensions/vectortiles/tutorial.rst (hunk 7 above)
Comment (ale-cristofori) [reply doc/en/user/source/extensions/vectortiles/tutorial.rst:179]: Done, corrected with the right paths to the files

File: doc/en/user/source/extensions/vectortiles/tutorial.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (ale-cristofori) [reply doc/en/user/source/extensions/vectortiles/tutorial.rst]: Included the qualification statement.

File: doc/en/user/source/extensions/vectortiles/tutorial.rst (hunk 1 above)
Comment (bradh) [reply doc/en/user/source/extensions/vectortiles/tutorial.rst]: It still appears to bad in the review version.  Can you check you have everything committed?

**Review by bradh — COMMENTED** [review]
OK.

File: doc/en/user/source/extensions/vectortiles/tutorial.rst [hunk 9]
```diff
@@ -80,18 +80,17 @@ For this tutorial, we'll be using the layer ``opengeo:countries`` to show off th
...
-Create OpenLayers application
------------------------------
+Create OpenLayers application - TMS Vector Tiles
+------------------------------------------------
 
-#. Create a ``www/vectortiles`` directory inside your GeoServer Data Directory.
+#. Create a ``www/tms-vectortiles`` directory inside your GeoServer Data Directory.
 
-#. Download the `latest version of OpenLayers <http://openlayers.org/download/>`_.
+#. Download the `latest version of OpenLayers <http://openlayers.org/download/>. Download the v<ol-version>-package.zip file`_.
```
Comment (aaime) [inline doc/en/user/source/extensions/vectortiles/tutorial.rst:88]: The docs build fails with:

```
07:36:54,213 [INFO]      [exec] /home/runner/work/geoserver/geoserver/doc/en/user/source/extensions/vectortiles/tutorial.rst:88: ERROR: Unknown target name: "latest version of openlayers <http://openlayers.org/download/>. download the v<ol-version>-package.zip file"."
```

The link seems to be missing a small bit at  the end, hopefully this is going to fix it.

```suggestion
#. Download the `latest version of OpenLayers <http://openlayers.org/download/>`_. Download the v<ol-version>-package.zip file`_.
```

Comment (ale-cristofori) [reply doc/en/user/source/extensions/vectortiles/tutorial.rst:88]: Thanks
